src/froll.c

```diff
@@ -397,6 +397,277 @@ void frollsumExact(double *x, uint64_t nx, ans_t *ans, int k, double fill, bool
+  }
+  if (verbose)
+    snprintf(end(ans->message[0]), 500, _("%s: processing algo %u took %.3fs\n"), __func__, algo, omp_get_wtime()-tic);
+}
+inline void windowmax(double *x, uint64_t o, int k, /* max of current window */ double *w, /* index of w within window */ uint64_t *iw) {
+  for (int i=0; i<k-1; i++) {
+    //Rprintf("windowmax iteration %d, offset %d, first x val %f, testing x[o+i-k+1] >= w[0]: x[%d-%d+1] >= w[0]: %f >= %f: %d\n", i, o, x[o], i, k, x[o+i-k+1], w[0], x[o+i-k+1] >= w[0]);
+    if (x[o+i-k+1] >= w[0]) { // what if that is never satisfied? test!
+      iw[0] = o+i-k+1;
+      w[0] = x[iw[0]];
+    }
+  }
+}
+inline void windowmaxnarm(double *x, uint64_t o, int k, bool narm, /* NA counter */ int *nc, double *w, uint64_t *iw) {
+  for (int i=0; i<k-1; i++) {
+    //Rprintf("windowmax iteration %d, offset %d, first x val %f, testing x[o+i-k+1] >= w[0]: x[%d-%d+1] >= w[0]: %f >= %f: %d\n", i, o, x[o], i, k, x[o+i-k+1], w[0], x[o+i-k+1] >= w[0]);
+    if (R_FINITE(x[o+i-k+1])) {
+      if (x[o+i-k+1] >= w[0]) { // what if that is never satisfied? test!
+        iw[0] = o+i-k+1;
+        w[0] = x[iw[0]];
+      }
+    } else if (narm) {
+      nc[0]++;
+    } else {
+      w[0] = NA_REAL;
+    }
+  }
+}
+/* fast rolling max - fast
+ * fast online algorithm do single pass over elements keeping track of recent max and its index
+ * if index of max is within progressing window then it keeps running single pass
+ * whenever max is leaving the window (index of max is outside of iterator minus window size) then new maximum is computed via nested loop on current complete window
+ * new max is used to continue outer single pass as long as new max index is not leaving the running window
+ * should scale well for bigger window size, may carry overhead for small window, needs benchmarking
+ * TODO NA handling
+ */
+void frollmaxFast(double *x, uint64_t nx, ans_t *ans, int k, double fill, bool narm, int hasna, bool verbose) {
+  if (verbose)
+    snprintf(end(ans->message[0]), 500, _("%s: running for input length %"PRIu64", window %d, hasna %d, narm %d\n"), "frollmaxFast", (uint64_t)nx, k, hasna, (int)narm);
+  double w = R_NegInf; // window max
+  uint64_t cmax = 0; // counter of nested loops for verbose
+  uint64_t iw = 0; // index of window max
+  bool truehasna = hasna>0;
+  if (!truehasna) {
+    int i;
+    for (i=0; i<k-1; i++) { // #loop_counter_not_local_scope_ok
+      if (!R_FINITE(x[i])) {
+        truehasna = true;
+        break;
+      }
+      if (x[i] >= w) { // >= rather than > because we track most recent maximum using iw
+        iw = i;
+        w = x[iw];
+      }
+      ans->dbl_v[i] = fill;
+    }
+    if (!truehasna && !R_FINITE(x[i])) {
+      if (x[i] >= w) {
+        iw = i;
+        w = x[iw];
+      } else {
+        // ensure iw ok, case when all window was -Inf or NA?
+      }
+      ans->dbl_v[i] = w;
+    } else {
+      truehasna = true;
+    }
+    if (!truehasna) {
+      for (uint64_t i=k; i<nx; i++) {
```

Comment:
why is `i` `uint64_t` here but `int` above?

Reply:
Because `k` is set in R as `integer`, whereas `x` could be a long vector, right? Probably worth a comment.

Reply:
Yes, k must not be bigger than int32 and x length can be int64

Reply:
I worry that "saving" 8 bytes on `uint64_t` for part of the loop here is over-optimizing at the expense of readability. Yes, we _can_ use `int` for the first `k` elements, but using a consistent storage type will prevent confusion/needing to know R implementation details like this.

In other words the cognitive load of using `int` is extremely high IMO. And "saving" is in quotes because it's more like "reducing RAM load by a tiny bit for a tiny fraction of time before the heavier loop kicks in".

Reply:
The purpose was that the code to be reflecting what is precisely happening rather than saving memory. For understanding the code I think it is better to keep it int when int is the type really used/needed. I can imagine it may be additionally confusing when we start to use data types that are not matching use case. So the purpose is rather for the code to be self descriptive rather than a memory optimization.

Reply:
I don't find the code self-descriptive.

> int is the type really used/needed

How so? It's used two ways: (1) index a long array. (2) RHS of assignment to a `uint64_t` variable (`iw  = i`), e.g.

https://github.com/Rdatatable/data.table/pull/5889/files#diff-6eb444de657f71e8b89b3d686c4ec70e4ecff4e33154481d0b1b416d09449f1eR477

So it seems to me `uint64_t` is the type really used/needed.

> I can imagine it may be additionally confusing when we start to use data types that are not matching use case.

Not sure I follow, but if you're really married to `int`, please add a comment (inside `frollmax`)

Reply:
`iw` is `uint64_t` so we don't need to re-declare it again. with `i` we declare it outside of the `for()` because it is needed after the loop finishes, and setting that to int64 upfront could give false impression that can be go over int64 range. I will add comments to explain that.